src/training_algorithms/sgd_training_algorithm.cpp

```diff
@@ -0,0 +1,286 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2016, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "lbann/training_algorithms/sgd_training_algorithm.hpp"
+#include "lbann/models/model.hpp"
+#include "lbann/callbacks/callback.hpp"
+
+namespace lbann {
+
+////////////////////////////////////////////////////////////
+// Evaluation and training
+////////////////////////////////////////////////////////////
+
+void sgd_training_algorithm::apply(execution_context& context,
+                                   model& model,
+                                   execution_mode mode,
+                                   termination_criteria const& term_criteria) {
+  sgd_execution_context& sgd_context = static_cast<sgd_execution_context&>(context);
+  const sgd_termination_criteria& sgd_term = static_cast<const sgd_termination_criteria&>(term_criteria);
+  switch(mode) {
```

Comment:
`sgd_training_algorithm` is basically implementing two algorithms based on the execution mode: SGD training or evaluation. We should split this functionality into `sgd_training_algorithm` and `evaluation_algorithm`.

Reply:
I agree that this repeated switch should be replaced with polymorphism. However, in the interest of "straight refactoring", I'm ok delaying this change to a future PR.